---
# Please do not edit this file directly; it is auto generated.
# Instead, please edit 03-DESeq.md in _episodes_rmd/
source: Rmd
title: "Differential Expression Analysis"
teaching: 40
exercises: 10
questions:
- "What transcriptomic changes we observe in mouse models carrying AD-related mutations?"
objectives:
- "Read in a count matrix and metadata."
- "Understand the data from AD mouse models"
- "Format the data for differential analysis"
- "Perform differential analysis using DESeq2."
- "Pathway enrichment of differentially expressed genes"
- "Save data for next lessons"
---



Author: Ravi Pandey, Jackson Laboratory


## LOAD libraries

~~~
suppressPackageStartupMessages(library("DESeq2"))
suppressPackageStartupMessages(library("ggplot2"))
suppressPackageStartupMessages(library("AnnotationDbi"))
suppressPackageStartupMessages(library("org.Mm.eg.db"))
suppressPackageStartupMessages(library("GO.db"))
suppressPackageStartupMessages(library("EnhancedVolcano"))
suppressPackageStartupMessages(library(tidyverse))
suppressPackageStartupMessages(library(dplyr))
suppressPackageStartupMessages(library(clusterProfiler))
~~~
{: .language-r}



## Reading Gene Expression Count matrix from  Previous Lesson
In this lesson, we will use the raw counts matrix and metadata downloaded in the previous lesson and will perform differential expression analysis. 

### RNA-Seq data from 5xFAD mouse models


~~~
counts <- read.delim("../data/htseqcounts_5XFAD.txt", check.names = FALSE)
~~~
{: .language-r}



~~~
Warning in file(file, "rt"): cannot open file '../data/htseqcounts_5XFAD.txt':
No such file or directory
~~~
{: .warning}



~~~
Error in file(file, "rt"): cannot open the connection
~~~
{: .error}

Reading Sample Metadata from Previous Lesson

~~~
covars <- readRDS("../data/covars_5XFAD.rds")
~~~
{: .language-r}



~~~
Warning in gzfile(file, "rb"): cannot open compressed file
'../data/covars_5XFAD.rds', probable reason 'No such file or directory'
~~~
{: .warning}



~~~
Error in gzfile(file, "rb"): cannot open the connection
~~~
{: .error}

Let's explore the data: 

Let’s look at the top of the metadata.

~~~
head(covars)
~~~
{: .language-r}



~~~
Error in h(simpleError(msg, call)): error in evaluating the argument 'x' in selecting a method for function 'head': object 'covars' not found
~~~
{: .error}

identify distinct groups using sample metadata

~~~
distinct(covars, sex, genotype, timepoint)
~~~
{: .language-r}



~~~
Error in eval(expr, envir, enclos): object 'covars' not found
~~~
{: .error}

We're going to explore the data further using a series of Challenges. 
You will be asked to look at the contents of some of the columns to see 
how the data are 
distributed.

> ## Challenge 1
> How many mice were used to produce this data? 
>
> > ## Solution to Challenge 1
> >
> > covars %>% group_by(sex,genotype,timepoint) %>% count()
> > dplyr::count(metadata, sex, genotype,timepoint)
> {: .solution}
{: .challenge}


How many rows and columns are there in `counts`?

~~~
dim(counts)
~~~
{: .language-r}



~~~
NULL
~~~
{: .output}

In the counts matrix, genes are in rows and samples are in columns. Let’s look at the first few rows.


~~~
head(counts,n=5)
~~~
{: .language-r}



~~~
                                                                                      
1 new("standardGeneric", .Data = function (object, ...)                               
2 standardGeneric("counts"), generic = structure("counts", package = "BiocGenerics"), 
3     package = "BiocGenerics", group = list(), valueClass = character(0),            
4     signature = "object", default = NULL, skeleton = (function (object,             
5         ...)                                                                        
~~~
{: .output}

As you can see, the gene ids are ENSEBL IDs. There is some risk that these may not be unique. Let’s check whether any of the gene symbols are duplicated. We will sum the number of duplicated gene symbols.


~~~
sum(duplicated(rownames(counts)))
~~~
{: .language-r}



~~~
[1] 0
~~~
{: .output}

The sum equals zero, so there are no duplicated gene symbols, which is good. Similarly, samples should be unique. Once again, let's verify this:


~~~
sum(duplicated(colnames(counts)))
~~~
{: .language-r}



~~~
[1] 0
~~~
{: .output}


### Formatting the count matrix
Now, as we see that gene_id is in first column of count matrix, but we will need only count data in matrix, so we will change the gene_id column to rownames.  
Converting the gene_id as rownames of count matrix

~~~
counts <- counts %>% column_to_rownames(.,var="gene_id") %>% as.data.frame()
~~~
{: .language-r}



~~~
Error in (function (cond) : error in evaluating the argument 'x' in selecting a method for function 'as.data.frame': is.data.frame(.data) is not TRUE
~~~
{: .error}

let's confirm if change is done correctly

~~~
head(counts,n=5)
~~~
{: .language-r}



~~~
                                                                                      
1 new("standardGeneric", .Data = function (object, ...)                               
2 standardGeneric("counts"), generic = structure("counts", package = "BiocGenerics"), 
3     package = "BiocGenerics", group = list(), valueClass = character(0),            
4     signature = "object", default = NULL, skeleton = (function (object,             
5         ...)                                                                        
~~~
{: .output}

As you can see from count table there are some genes that start with **"ENSG"** and others start with **"ENSMUSG"**. **"ENSG"** referes to human gene ENSEMBL id and **"ENSMUSG"** refer to mouse ENSEMBL id. Let's check how many gene_ids are NOT from the mouse genome by searching for the string "MUS" (as in Mus musculus) in the rownames of count matrix

~~~
counts[,1:6] %>% 
  filter(!str_detect(rownames(.), "MUS"))
~~~
{: .language-r}



~~~
Error in counts[, 1:6]: object of type 'closure' is not subsettable
~~~
{: .error}
Ok, so we see there are two human genes in out count matrix. Why? What genes are they?


Briefly, 5xFAD mouse strain harbors two human transgenes APP ("ENSG00000142192") and PSEN1 ("ENSG00000080815") and inserted into exon 2 of the mouse Thy1 gene. To validate 5XFAD strain and capture expression of human transgene APP and PS1, a custom mouse genomic sequences was created and we quantified expression of human as well as mouse App ("ENSMUSG00000022892") and Psen1 ("ENSMUSG00000019969") genes by our MODEL-AD RNA-Seq pipeline.

### Validation of 5xFAD mouse strain

#First we convert the dataframe to longer format and join our covariates by MouseID

~~~
count_tpose <- counts  %>%
                rownames_to_column(.,var="gene_id") %>% 
                filter(gene_id %in% c("ENSG00000080815","ENSMUSG00000019969","ENSG00000142192","ENSMUSG00000022892")) %>% 
                pivot_longer(.,cols = -"gene_id",names_to = "specimenID",values_to="counts") %>% as.data.frame() %>%
                left_join(covars ,by="specimenID") %>% as.data.frame()
~~~
{: .language-r}



~~~
Error in h(simpleError(msg, call)): error in evaluating the argument 'x' in selecting a method for function 'as.data.frame': error in evaluating the argument 'x' in selecting a method for function 'as.data.frame': is.data.frame(df) is not TRUE
~~~
{: .error}



~~~
head(count_tpose) 
~~~
{: .language-r}



~~~
Error in h(simpleError(msg, call)): error in evaluating the argument 'x' in selecting a method for function 'head': object 'count_tpose' not found
~~~
{: .error}



~~~
#make the age column a factor and re-order the levels
count_tpose$timepoint <- factor(count_tpose$timepoint,levels=c("4 mo","6 mo","12 mo"))
~~~
{: .language-r}



~~~
Error in eval(expr, envir, enclos): object 'count_tpose' not found
~~~
{: .error}



~~~
# rename the gene id to gene symbol
count_tpose$gene_id[count_tpose$gene_id %in% "ENSG00000142192"] <- "Human APP"
~~~
{: .language-r}



~~~
Error in eval(ei, envir): object 'count_tpose' not found
~~~
{: .error}



~~~
count_tpose$gene_id[count_tpose$gene_id %in% "ENSG00000080815"] <- "Human PSEN1"
~~~
{: .language-r}



~~~
Error in eval(ei, envir): object 'count_tpose' not found
~~~
{: .error}



~~~
count_tpose$gene_id[count_tpose$gene_id %in% "ENSMUSG00000022892"] <- "Mouse App"
~~~
{: .language-r}



~~~
Error in eval(ei, envir): object 'count_tpose' not found
~~~
{: .error}



~~~
count_tpose$gene_id[count_tpose$gene_id %in% "ENSMUSG00000019969"] <- "Mouse Psen1"
~~~
{: .language-r}



~~~
Error in eval(ei, envir): object 'count_tpose' not found
~~~
{: .error}



~~~
#Create simple box plots showing normalized counts by genotype and time point, faceted by sex.
count_tpose %>% 
  ggplot(aes(x=timepoint, y=counts, color=genotype)) +
  geom_boxplot() + 
  geom_point(position=position_jitterdodge()) +
  facet_wrap(~sex+gene_id) +theme_bw()
~~~
{: .language-r}



~~~
Error in eval(expr, envir, enclos): object 'count_tpose' not found
~~~
{: .error}
You will notice expression of Human APP is higher in 5XFAD carriers but lower in non-carriers. However mouse App expressed in both 5XFAD carrier and non-carrier. 

We are going to sum the counts from both ortholgous genes (human APP and mouse App; human PSEN1 and mouse Psen1) and save summed expression as expression of mouse genes, respectively to match with gene names in control mice.

~~~
#merge mouse and human APP gene raw count
counts[rownames(counts) %in% "ENSMUSG00000022892",] <- counts[rownames(counts) %in% "ENSMUSG00000022892",] + counts[rownames(counts) %in% "ENSG00000142192",]
~~~
{: .language-r}



~~~
Error in counts[rownames(counts) %in% "ENSMUSG00000022892", ]: object of type 'closure' is not subsettable
~~~
{: .error}



~~~
counts <- counts[!rownames(counts) %in% c("ENSG00000142192"),]
~~~
{: .language-r}



~~~
Error in counts[!rownames(counts) %in% c("ENSG00000142192"), ]: object of type 'closure' is not subsettable
~~~
{: .error}



~~~
#merge mouse and human PS1 gene raw count
counts[rownames(counts) %in% "ENSMUSG00000019969",] <- counts[rownames(counts) %in% "ENSMUSG00000019969",] + counts[rownames(counts) %in% "ENSG00000080815",]
~~~
{: .language-r}



~~~
Error in counts[rownames(counts) %in% "ENSMUSG00000019969", ]: object of type 'closure' is not subsettable
~~~
{: .error}



~~~
counts <- counts[!rownames(counts) %in% c("ENSG00000080815"),]
~~~
{: .language-r}



~~~
Error in counts[!rownames(counts) %in% c("ENSG00000080815"), ]: object of type 'closure' is not subsettable
~~~
{: .error}

Let's verify if expression of both human genes have been merged or not:

~~~
counts[,1:6] %>% 
  filter(!str_detect(rownames(.), "MUS"))
~~~
{: .language-r}



~~~
Error in counts[, 1:6]: object of type 'closure' is not subsettable
~~~
{: .error}


What proportion of genes have zero counts in all samples?

~~~
gene_sums <- data.frame(gene_id = rownames(counts),
                        sums    = Matrix::rowSums(counts))
~~~
{: .language-r}



~~~
Error in base::rowSums(x, na.rm = na.rm, dims = dims, ...): 'x' must be an array of at least two dimensions
~~~
{: .error}



~~~
sum(gene_sums$sums == 0)
~~~
{: .language-r}



~~~
Error in eval(expr, envir, enclos): object 'gene_sums' not found
~~~
{: .error}
We can see that 9691 (17%) genes have no reads at all associated with them. In the next lesson, we will remove genes that have no counts in any samples.

## Differential Analysis using DESeq2

Now, after exploring and formatting the data, We will look for differential expression between the control and 5xFAD mice at different ages for both sexes. The differentially expressed genes (DEGs) can inform our understanding of how the 5XFAD mutation affect the biological processes.

DESeq2 analysis consist of multiple steps. We are going to briefly understand some of the important steps using subset of data and then we will perform differential analysis on whole dataset. For detailed analysis see the [DESeq2](http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html) tutorial.

First, order the data (so counts and metadata match) and save in another variable

~~~
rawdata <- counts[,sort(colnames(counts))]
~~~
{: .language-r}



~~~
Error in counts[, sort(colnames(counts))]: object of type 'closure' is not subsettable
~~~
{: .error}



~~~
metadata <- covars[sort(rownames(covars)),]
~~~
{: .language-r}



~~~
Error in eval(expr, envir, enclos): object 'covars' not found
~~~
{: .error}

subset the counts matrix and sample metadata to include only 12 month old male mice. You can amend the code to compare wild type and 5XFAD mice from either sex, at any time point.


~~~
meta.12M.Male <- metadata[(metadata$sex=="male" & metadata$timepoint=='12 mo'),]
~~~
{: .language-r}



~~~
Error in (function (cond) : error in evaluating the argument 'i' in selecting a method for function '[': object of type 'closure' is not subsettable
~~~
{: .error}



~~~
meta.12M.Male
~~~
{: .language-r}



~~~
Error in eval(expr, envir, enclos): object 'meta.12M.Male' not found
~~~
{: .error}



~~~
dat <- as.matrix(rawdata[,colnames(rawdata) %in% rownames(meta.12M.Male)])
~~~
{: .language-r}



~~~
Error in h(simpleError(msg, call)): error in evaluating the argument 'x' in selecting a method for function 'as.matrix': object 'rawdata' not found
~~~
{: .error}



~~~
colnames(dat)
~~~
{: .language-r}



~~~
Error in h(simpleError(msg, call)): error in evaluating the argument 'x' in selecting a method for function 'colnames': object 'dat' not found
~~~
{: .error}



~~~
rownames(meta.12M.Male)
~~~
{: .language-r}



~~~
Error in h(simpleError(msg, call)): error in evaluating the argument 'x' in selecting a method for function 'rownames': object 'meta.12M.Male' not found
~~~
{: .error}



~~~
match(colnames(dat),rownames(meta.12M.Male))
~~~
{: .language-r}



~~~
Error in h(simpleError(msg, call)): error in evaluating the argument 'x' in selecting a method for function 'match': error in evaluating the argument 'x' in selecting a method for function 'colnames': object 'dat' not found
~~~
{: .error}


Next, we build the DESeqDataSet using the following function:

~~~
ddsHTSeq <- DESeqDataSetFromMatrix(countData=dat, 
                                   colData=meta.12M.Male, 
                                   design = ~genotype)
~~~
{: .language-r}



~~~
Error in h(simpleError(msg, call)): error in evaluating the argument 'x' in selecting a method for function 'ncol': object 'dat' not found
~~~
{: .error}



~~~
ddsHTSeq
~~~
{: .language-r}



~~~
Error in eval(expr, envir, enclos): object 'ddsHTSeq' not found
~~~
{: .error}

### Pre-filtering
While it is not necessary to pre-filter low count genes before running the DESeq2 functions, there are two reasons which make pre-filtering useful: by removing rows in which there are very few reads, we reduce the memory size of the dds data object, and we increase the speed of the transformation and testing functions within DESeq2. It can also improve visualizations, as features with no information for differential expression are not plotted.

Here we perform a minimal pre-filtering to keep only rows that have at least 10 reads total. 

~~~
ddsHTSeq <- ddsHTSeq[rowSums(counts(ddsHTSeq)) >= 10,]
~~~
{: .language-r}



~~~
Error in eval(expr, envir, enclos): object 'ddsHTSeq' not found
~~~
{: .error}



~~~
ddsHTSeq
~~~
{: .language-r}



~~~
Error in eval(expr, envir, enclos): object 'ddsHTSeq' not found
~~~
{: .error}

### Reference level
By default, R will choose a reference level for factors based on alphabetical order. Then, if you never tell the DESeq2 functions which level you want to compare against (e.g. which level represents the control group), the comparisons will be based on the alphabetical order of the levels.

specifying the reference-level to **5XFAD_noncarrier**:

~~~
ddsHTSeq$genotype <- relevel(ddsHTSeq$genotype,ref="5XFAD_noncarrier")  
~~~
{: .language-r}



~~~
Error in eval(expr, envir, enclos): object 'ddsHTSeq' not found
~~~
{: .error}

Run the standard differential expression analysis steps that is wrapped into a single function, `DESeq`.

~~~
dds <- DESeq(ddsHTSeq,parallel = TRUE)
~~~
{: .language-r}



~~~
Error in eval(expr, envir, enclos): object 'ddsHTSeq' not found
~~~
{: .error}

Results tables are generated using the function `results`, which extracts a results table with log2 fold changes, p values and adjusted p values. By default the argument alpha is set to 0.1. If the adjusted p value cutoff will be a value other than 0.1, alpha should be set to that value:


~~~
res <- results(dds,alpha=0.05)  # setting 0.05 as significant threshold
~~~
{: .language-r}



~~~
Error in eval(expr, envir, enclos): object 'dds' not found
~~~
{: .error}



~~~
res
~~~
{: .language-r}



~~~
Error in eval(expr, envir, enclos): object 'res' not found
~~~
{: .error}

We can order our results table by the smallest p value:

~~~
resOrdered <- res[order(res$pvalue),]
~~~
{: .language-r}



~~~
Error in eval(expr, envir, enclos): object 'res' not found
~~~
{: .error}



~~~
head(resOrdered,n=10)
~~~
{: .language-r}



~~~
Error in h(simpleError(msg, call)): error in evaluating the argument 'x' in selecting a method for function 'head': object 'resOrdered' not found
~~~
{: .error}

we can summarize some basic tallies using the summary function.

~~~
summary(res)
~~~
{: .language-r}



~~~
Error in h(simpleError(msg, call)): error in evaluating the argument 'object' in selecting a method for function 'summary': object 'res' not found
~~~
{: .error}

How many adjusted p-values were less than 0.05?

~~~
sum(res$padj < 0.05, na.rm=TRUE)
~~~
{: .language-r}



~~~
Error in eval(expr, envir, enclos): object 'res' not found
~~~
{: .error}


> ## Challenge 2
> How many adjusted p-values were less than 0.1? 
>
> > ## Solution to Challenge 2
> >
> > sum(res$padj < 0.1, na.rm=TRUE) 
> {: .solution}
{: .challenge}


### Function to convert ensembleIDs to common gene names
We'll use a package to translate mouse ENSEMBL IDS to gene names. Run this function and they will be called up when assembling results from the differential expression analysis


~~~
map_function.df <- function(x, inputtype, outputtype) {
  mapIds(
    org.Mm.eg.db,
    keys = row.names(x),
    column = outputtype,
    keytype = inputtype,
    multiVals = "first"
  )
}
~~~
{: .language-r}

### Generating Result table

~~~
All_res <- as.data.frame(res) %>% 
  mutate(symbol = map_function.df(res,"ENSEMBL","SYMBOL")) %>%    ##run map_function to add symbol of gene corresponding to ENSEBL ID
  mutate(EntrezGene = map_function.df(res,"ENSEMBL","ENTREZID")) %>%  ##run map_function to add Entrez ID of gene corresponding to ENSEBL ID
  dplyr::select("symbol", "EntrezGene","baseMean", "log2FoldChange", "lfcSE", "stat", "pvalue", "padj")
~~~
{: .language-r}



~~~
Error in h(simpleError(msg, call)): error in evaluating the argument 'x' in selecting a method for function 'as.data.frame': object 'res' not found
~~~
{: .error}



~~~
head(All_res)
~~~
{: .language-r}



~~~
Error in h(simpleError(msg, call)): error in evaluating the argument 'x' in selecting a method for function 'head': object 'All_res' not found
~~~
{: .error}

### Extracting genes that are significantly expressed


~~~
dseq_res <- subset(All_res[order(All_res$padj), ], padj < 0.05)
~~~
{: .language-r}



~~~
Error in h(simpleError(msg, call)): error in evaluating the argument 'x' in selecting a method for function 'subset': object 'All_res' not found
~~~
{: .error}
Wow! We have a lot of genes with apparently very strong statistically significant differences between the control and 5xFAD carrier.


~~~
dim(dseq_res)
~~~
{: .language-r}



~~~
Error in eval(expr, envir, enclos): object 'dseq_res' not found
~~~
{: .error}



~~~
head(dseq_res)
~~~
{: .language-r}



~~~
Error in h(simpleError(msg, call)): error in evaluating the argument 'x' in selecting a method for function 'head': object 'dseq_res' not found
~~~
{: .error}

## Exploring and exporting results

#### Exporting results to CSV files
we can save results file into a csv file like this:

~~~
write.csv(All_res,file="../result/All_5xFAD_12months_male.csv")
~~~
{: .language-r}



~~~
Error in eval(expr, p): object 'All_res' not found
~~~
{: .error}



~~~
write.csv(dseq_res,file="../result/DEG_5xFAD_12months_male.csv")
~~~
{: .language-r}



~~~
Error in eval(expr, p): object 'dseq_res' not found
~~~
{: .error}

## Volcano plot

We can visualize the differential expression results using Volcano plot function from **EnhancedVolcano** package.
For the most basic volcano plot, only a single data-frame, data-matrix, or tibble of test results is required, containing point labels, log2FC, and adjusted or unadjusted P values. The default cut-off for log2FC is >|2|; the default cut-off for P value is 10e-6.

~~~
EnhancedVolcano(All_res,
                                   lab = (All_res$symbol),
                                   x = 'log2FoldChange',
                                   y = 'padj',legendPosition = 'none',
                                   title = 'Volcano plot:Differential Expression Results',
                                   subtitle = '',
                                   FCcutoff = 0.1,
                                   pCutoff = 0.05,
                                   xlim = c(-3, 6))
~~~
{: .language-r}



~~~
Error in eval(expr, envir, enclos): object 'All_res' not found
~~~
{: .error}
You can see that some top significantly expressed are immune/inflammation-related genes such as Ctsd, C4b, Csf1 etc. These genes are upregulated in the 5XFAD strain. 

## Principal component plot of the samples

~~~
ddsHTSeq <- DESeqDataSetFromMatrix(countData=as.matrix(rawdata), colData=metadata, design= ~ genotype)
~~~
{: .language-r}



~~~
Error in h(simpleError(msg, call)): error in evaluating the argument 'x' in selecting a method for function 'ncol': error in evaluating the argument 'x' in selecting a method for function 'as.matrix': object 'rawdata' not found
~~~
{: .error}



~~~
ddsHTSeq <- ddsHTSeq[rowSums(counts(ddsHTSeq)>1) >= 10, ]
~~~
{: .language-r}



~~~
Error in eval(expr, envir, enclos): object 'ddsHTSeq' not found
~~~
{: .error}



~~~
dds <- DESeq(ddsHTSeq,parallel = TRUE)
~~~
{: .language-r}



~~~
Error in eval(expr, envir, enclos): object 'ddsHTSeq' not found
~~~
{: .error}



~~~
vsd <- varianceStabilizingTransformation(dds, blind=FALSE)
~~~
{: .language-r}



~~~
Error in h(simpleError(msg, call)): error in evaluating the argument 'x' in selecting a method for function 'colnames': object 'dds' not found
~~~
{: .error}



~~~
plotPCA(vsd, intgroup=c("genotype", "sex","timepoint"))
~~~
{: .language-r}



~~~
Error in h(simpleError(msg, call)): error in evaluating the argument 'object' in selecting a method for function 'plotPCA': object 'vsd' not found
~~~
{: .error}

It is also possible to customize the PCA plot using the ggplot function.

~~~
pcaData <- plotPCA(vsd, intgroup=c("genotype", "sex","timepoint"), returnData=TRUE)
~~~
{: .language-r}



~~~
Error in h(simpleError(msg, call)): error in evaluating the argument 'object' in selecting a method for function 'plotPCA': object 'vsd' not found
~~~
{: .error}



~~~
percentVar <- round(100 * attr(pcaData, "percentVar"))
~~~
{: .language-r}



~~~
Error in eval(expr, envir, enclos): object 'pcaData' not found
~~~
{: .error}



~~~
ggplot(pcaData, aes(PC1, PC2,color=genotype, shape=sex)) + 
  geom_point(size=3) +
  geom_text(aes(label=timepoint),hjust=0.5, vjust=2,size =3.5) +
  labs(x= paste0("PC1: ",percentVar[1],"% variance"), y= paste0("PC2: ",percentVar[2],"% variance"))
~~~
{: .language-r}



~~~
Error in eval(expr, envir, enclos): object 'pcaData' not found
~~~
{: .error}

PCA identified genotype and sex being a major source of variation in between 5XFAD and WT mice. Female and male samples clustered distinctly at all ages, suggesting the presence of sex-biased molecular changes in animals. 

## Function for Differential analysis using DESeq2

Finally, we can built a function for differential analysis that consist of all above discussed steps. It will require to input sorted raw count matrix, sample metadata and define the reference group.

~~~
DEG <- function(rawdata,meta,
                   include.batch = FALSE,
                   ref = ref) {
  dseq_res <- data.frame()
  All_res <- data.frame()
  
  if (include.batch) {
    cat("Including batch as covariate\n")
    design_formula <- ~ Batch + genotype
  }
  else{
    design_formula <- ~ genotype
  }
  
  dat2 <- as.matrix(rawdata[, colnames(rawdata) %in% rownames(meta)])
  ddsHTSeq <-
    DESeqDataSetFromMatrix(countData = dat2,
                           colData = meta,
                           design = design_formula)
  ddsHTSeq <- ddsHTSeq[rowSums(counts(ddsHTSeq)) >= 10, ]
  ddsHTSeq$genotype <- relevel(ddsHTSeq$genotype, ref = ref)
  dds <- DESeq(ddsHTSeq, parallel = TRUE)
  
  res <- results(dds, alpha = 0.05)
  #summary(res)
  
  res$symbol <- map_function.df(res,"ENSEMBL","SYMBOL")
  
  res$EntrezGene <- map_function.df(res,"ENSEMBL","ENTREZID")
  
  All_res <<- as.data.frame(res[, c("symbol", "EntrezGene","baseMean", "log2FoldChange", "lfcSE", "stat", "pvalue", "padj")])

}
~~~
{: .language-r}

Let's use this function to analyze all groups present in our data.

## Differential Analysis of all groups

First, we add a **Group** column to our metadata table that will combine all variable of interest for that sample.

~~~
metadata$Group <- paste0(metadata$genotype,"-",metadata$sex,"-",metadata$timepoint)
~~~
{: .language-r}



~~~
Error in metadata$genotype: object of type 'closure' is not subsettable
~~~
{: .error}



~~~
unique(metadata$Group)
~~~
{: .language-r}



~~~
Error in (function (cond) : error in evaluating the argument 'x' in selecting a method for function 'unique': object of type 'closure' is not subsettable
~~~
{: .error}

Next, we create a comparison table that has all cases and controls that we would like to compare with each other. Here I have made comparison group for age and sex-matched 5xFAD carriers vs 5xFAD_noncarriers:

~~~
comparisons <-  data.frame(control=c("5XFAD_noncarrier-male-4 mo", "5XFAD_noncarrier-female-4 mo", "5XFAD_noncarrier-male-6 mo", 
                                     "5XFAD_noncarrier-female-6 mo","5XFAD_noncarrier-male-12 mo", "5XFAD_noncarrier-female-12 mo"), 
                           case=c("5XFAD_carrier-male-4 mo", "5XFAD_carrier-female-4 mo", "5XFAD_carrier-male-6 mo", 
                                  "5XFAD_carrier-female-6 mo","5XFAD_carrier-male-12 mo", "5XFAD_carrier-female-12 mo")
                           )
~~~
{: .language-r}


~~~
comparisons
~~~
{: .language-r}



~~~
                        control                       case
1    5XFAD_noncarrier-male-4 mo    5XFAD_carrier-male-4 mo
2  5XFAD_noncarrier-female-4 mo  5XFAD_carrier-female-4 mo
3    5XFAD_noncarrier-male-6 mo    5XFAD_carrier-male-6 mo
4  5XFAD_noncarrier-female-6 mo  5XFAD_carrier-female-6 mo
5   5XFAD_noncarrier-male-12 mo   5XFAD_carrier-male-12 mo
6 5XFAD_noncarrier-female-12 mo 5XFAD_carrier-female-12 mo
~~~
{: .output}


Finally, we implement our DEG function on each comparison and store the result table in a list and data frame:

~~~
# initiate a empty list and data frame to save results
DE_5xFAD.list <- list()
DE_5xFAD.df <- data.frame()

for (i in 1:nrow(comparisons))
  {
    meta <- metadata[metadata$Group %in% comparisons[i,],]
    DEG(rawdata,meta,ref = "5XFAD_noncarrier")
    
    #append results in data frame
    DE_5xFAD.df <- rbind(DE_5xFAD.df,All_res %>% mutate(model="5xFAD",sex=unique(meta$sex),age=unique(meta$timepoint)))
    
    #append results in list
    DE_5xFAD.list[[i]] <- All_res
    names(DE_5xFAD.list)[i] <- paste0(comparisons[i,2])
}
~~~
{: .language-r}



~~~
Error in h(simpleError(msg, call)): error in evaluating the argument 'i' in selecting a method for function '[': error in evaluating the argument 'x' in selecting a method for function '%in%': object of type 'closure' is not subsettable
~~~
{: .error}

Let's explore the result stored in our list:

~~~
names(DE_5xFAD.list)
~~~
{: .language-r}



~~~
NULL
~~~
{: .output}

We can easily extract result table for any group of interest by using **$** and name of group. Let's check top few rows from 5XFAD_carrier-male-4 mo group:

~~~
head(DE_5xFAD.list$`5XFAD_carrier-male-4 mo`)
~~~
{: .language-r}



~~~
NULL
~~~
{: .output}

Let's check the result stored as dataframe:

~~~
head(DE_5xFAD.df)
~~~
{: .language-r}



~~~
data frame with 0 columns and 0 rows
~~~
{: .output}

Check if result is present for all ages:

~~~
unique((DE_5xFAD.df$age))
~~~
{: .language-r}



~~~
NULL
~~~
{: .output}

Check if result is present for both sexes:

~~~
unique((DE_5xFAD.df$sex))
~~~
{: .language-r}



~~~
NULL
~~~
{: .output}

Check number of genes in each group:

~~~
count(DE_5xFAD.df,model,sex,age)
~~~
{: .language-r}



~~~
Error in `count()`:
! Must group by variables found in `.data`.
Column `model` is not found.
Column `sex` is not found.
Column `age` is not found.
~~~
{: .error}

#### Check number of genes significantly differentially expressed in all cases compared to age and sex-matched controls:

~~~
degs.up <- map(DE_5xFAD.list, ~length(which(.x$padj<0.05 & .x$log2FoldChange>0)))
degs.down <- map(DE_5xFAD.list, ~length(which(.x$padj<0.05 & .x$log2FoldChange<0)))
deg <- data.frame(Cases=names(degs.up), Up_DEGs.pval.05=as.vector(unlist(degs.up)),Down_DEGs.pval.05=as.vector(unlist(degs.down)))
knitr::kable(deg)
~~~
{: .language-r}

Interestingly, in females more genes are differentially expressed at 4 months and with age more genes are differentially expressed. Male mice is catching up female mice at later time-point.

## Pathway Enrichment
We may wish to look for enrichment of biological pathways in a list of differentially expressed genes. Here we will test for enrichment of KEGG pathways using  using enrichKEGG function in [clusterProfiler](https://bioconductor.org/packages/release/bioc/vignettes/clusterProfiler/inst/doc/clusterProfiler.html) package.

~~~
dat <- list(FAD_M_4=subset(DE_5xFAD.list$`5XFAD_carrier-male-4 mo`[order(DE_5xFAD.list$`5XFAD_carrier-male-4 mo`$padj), ], padj < 0.05) %>% pull(EntrezGene),
            FAD_F_4=subset(DE_5xFAD.list$`5XFAD_carrier-female-4 mo`[order(DE_5xFAD.list$`5XFAD_carrier-female-4 mo`$padj), ], padj < 0.05) %>% pull(EntrezGene))
~~~
{: .language-r}



~~~
Error in h(simpleError(msg, call)): error in evaluating the argument 'x' in selecting a method for function 'subset': argument 1 is not a vector
~~~
{: .error}



~~~
## perform enrichment analysis
enrich_pathway <- compareCluster(dat,
                                 fun = "enrichKEGG",
                                  pvalueCutoff = 0.05,
                                  organism = "mmu"
                                 )
~~~
{: .language-r}



~~~
Error in eval(expr, envir, enclos): object 'dat' not found
~~~
{: .error}



~~~
enrich_pathway@compareClusterResult$Description <- gsub(" - Mus musculus \\(house mouse)","",enrich_pathway@compareClusterResult$Description)
~~~
{: .language-r}



~~~
Error in h(simpleError(msg, call)): error in evaluating the argument 'x' in selecting a method for function 'gsub': object 'enrich_pathway' not found
~~~
{: .error}

Let's plot top enriched functions using dotplot function of clusterProfiler package.

~~~
clusterProfiler::dotplot(enrich_pathway,showCategory=10,font.size = 14,title="Enriched KEGG Pathways")
~~~
{: .language-r}



~~~
Error in h(simpleError(msg, call)): error in evaluating the argument 'object' in selecting a method for function 'dotplot': object 'enrich_pathway' not found
~~~
{: .error}

What does this plot infer?

## Save Data for Next Lesson 

We will use the results data in the next lesson. Save it now and we will load it at the beginning of the next lesson. We will use R’s save command to [save]((https://stat.ethz.ch/R-manual/R-devel/library/base/html/save.html)) the objects in compressed, binary format. The `save` command is useful when you want to save multiple objects in one file.

~~~
save(DE_5xFAD.df,DE_5xFAD.list,file="../result/DEAnalysis_5XFAD.Rdata")
~~~
{: .language-r}



~~~
Warning in gzfile(file, "wb"): cannot open compressed file
'../result/DEAnalysis_5XFAD.Rdata', probable reason 'No such file or directory'
~~~
{: .warning}



~~~
Error in gzfile(file, "wb"): cannot open the connection
~~~
{: .error}


> ## Challenge 3
> ## Challenge 3
> Draw volcano plot for 6 months old female 5xFAD_carrier ? 
>
> > ## Solution to Challenge 3
> >
> > EnhancedVolcano(DE_5xFAD.list$`5XFAD_carrier-female-6 mo`,
                                   lab = (DE_5xFAD.list$`5XFAD_carrier-female-6 mo`$symbol),
                                   x = 'log2FoldChange',
                                   y = 'padj',legendPosition = 'none',
                                   title = 'Volcano plot:Differential Expression Results',
                                   subtitle = '',
                                   FCcutoff = 0.1,
                                   pCutoff = 0.05,
                                   xlim = c(-5, 5))
> {: .solution}
{: .challenge}


### Loading RNA-Seq data from LOAD1 (APOE4*Trem2 Mouse Model)
There is another RNA-Seq data from LOAD1 mouse cohort. This cohort has three distinct mice strains: 
*APOE4: humanized Apoe knock-in strain, mouse Apoe allele was replaced by human APOE gene sequence; 
*Trem2.R47H:knock out mutant of the Trem2, R47H point mutation with two silent mutation and 
*APOE4.Trem2.R47H (LOAD1): double mutant strain carries humanized APOE (isoform 4) knock in mutation and R47H point mutation of the Trem2 gene. 

Let's read the count data and sample metadata as we did for 5XFAD samples.

~~~
counts <- read.delim("../data/htseqcounts_APTR.txt", check.names = FALSE)
~~~
{: .language-r}



~~~
Error in file(file, "rt"): cannot open the connection
~~~
{: .error}


~~~
ind_meta <- read.csv("../data/metadata_RNASEQ_ALLJAX_402Samples.csv") 
~~~
{: .language-r}



~~~
Warning in file(file, "rt"): cannot open file
'../data/metadata_RNASEQ_ALLJAX_402Samples.csv': No such file or directory
~~~
{: .warning}



~~~
Error in file(file, "rt"): cannot open the connection
~~~
{: .error}
 
Let’s look at the first few rows.
 

~~~
head(ind_meta)
~~~
{: .language-r}



~~~
Error in h(simpleError(msg, call)): error in evaluating the argument 'x' in selecting a method for function 'head': object 'ind_meta' not found
~~~
{: .error}
Let's quickly explore the sample metadata:

~~~
dim(ind_meta)
~~~
{: .language-r}



~~~
Error in eval(expr, envir, enclos): object 'ind_meta' not found
~~~
{: .error}

How many strains are in this sample metadata?

~~~
table(ind_meta$Genotype)
~~~
{: .language-r}



~~~
Error in eval(expr, envir, enclos): object 'ind_meta' not found
~~~
{: .error}

How many samples are in each genotype for each sex and age group? 

~~~
ind_meta %>% group_by(Sex,Genotype,Age) %>% count()
~~~
{: .language-r}



~~~
Error in eval(expr, envir, enclos): object 'ind_meta' not found
~~~
{: .error}

We notice that there is a column **Batch** in sample metadata. 

Let's see how many samples are in each batch.

~~~
table(ind_meta$Batch)
~~~
{: .language-r}



~~~
Error in eval(expr, envir, enclos): object 'ind_meta' not found
~~~
{: .error}
So, total 5 batches. In bulk RNA-Seq experiments, it is usually vital that we apply a correction for samples profiled in different batches. Due to tim-constraint we do not want to do it in this lesson. So, we will use samples from only one batch in this lesson. 

Let's check which genotype are in which batch?

~~~
table(ind_meta$Batch,ind_meta$Genotype)
~~~
{: .language-r}



~~~
Error in eval(expr, envir, enclos): object 'ind_meta' not found
~~~
{: .error}
We are going to use samples from Batch 1 as it has more samples and contain most of our LOAD1 cohort samples that we are interested. 

Subset sample metadata for batch 1 only and Convert Mouse ID column to rownames.

~~~
covar <- ind_meta %>% filter(Batch==1) %>% remove_rownames() %>% column_to_rownames(.,var="Mouse.ID")  
~~~
{: .language-r}



~~~
Error in eval(expr, envir, enclos): object 'ind_meta' not found
~~~
{: .error}

How many rows and columns are there in `covar`?

~~~
dim(covar)
~~~
{: .language-r}



~~~
Error in eval(expr, envir, enclos): object 'covar' not found
~~~
{: .error}

Let’s look at the first few rows.

~~~
head(covar)
~~~
{: .language-r}



~~~
Error in h(simpleError(msg, call)): error in evaluating the argument 'x' in selecting a method for function 'head': object 'covar' not found
~~~
{: .error}

Check mouse strains and are there numbers in `covar`?

~~~
table(covar$Genotype)
~~~
{: .language-r}



~~~
Error in eval(expr, envir, enclos): object 'covar' not found
~~~
{: .error}


> ## Challenge 4
> ## Challenge 4
> How many samples are in each genotype for each sex and age group? 
>
> > ## Solution to Challenge 4
> > covar %>% group_by(Sex,Genotype,Age) %>% count()
> > 
> {: .solution}
{: .challenge}

Let's explore count data.


~~~
head(counts,n=5)
~~~
{: .language-r}



~~~
                                                                                      
1 new("standardGeneric", .Data = function (object, ...)                               
2 standardGeneric("counts"), generic = structure("counts", package = "BiocGenerics"), 
3     package = "BiocGenerics", group = list(), valueClass = character(0),            
4     signature = "object", default = NULL, skeleton = (function (object,             
5         ...)                                                                        
~~~
{: .output}
Converting the gene_id as rownames of count matrix

~~~
counts <- counts %>% column_to_rownames(.,var="gene_ID") %>% as.data.frame()
~~~
{: .language-r}



~~~
Error in (function (cond) : error in evaluating the argument 'x' in selecting a method for function 'as.data.frame': is.data.frame(.data) is not TRUE
~~~
{: .error}

How many rows and columns are there in `counts`?

~~~
dim(counts)
~~~
{: .language-r}



~~~
NULL
~~~
{: .output}
In the counts matrix, genes are in rows and samples are in columns.

There are 140 samples in batch 1 as we saw in `covar` and samples in counts matrix is 234. So, we need to subset counts matrix for samples in `covar`. 

~~~
counts <- counts[,colnames(counts) %in% rownames(covar)]
~~~
{: .language-r}



~~~
Error in h(simpleError(msg, call)): error in evaluating the argument 'j' in selecting a method for function '[': error in evaluating the argument 'table' in selecting a method for function '%in%': error in evaluating the argument 'x' in selecting a method for function 'rownames': object 'covar' not found
~~~
{: .error}


~~~
dim(counts)
~~~
{: .language-r}



~~~
NULL
~~~
{: .output}
Now there are count data from 140 samples present in metadata.

Check how many gene_ids are NOT from the mouse genome by searching for the string "MUS" (as in Mus musculus) in the rownames of count matrix

~~~
counts[,1:6] %>% 
  filter(!str_detect(rownames(.), "MUS"))
~~~
{: .language-r}



~~~
Error in counts[, 1:6]: object of type 'closure' is not subsettable
~~~
{: .error}
We see entry of gene id "ENSG00000130203". This is human APOE gene that has been quantified by our MODEL-AD RNA-Seq pipeline.  

Here you can also check expression of mouse Apoe. 

~~~
counts[rownames(counts) %in% "ENSMUSG00000002985",1:10]
~~~
{: .language-r}



~~~
Error in counts[rownames(counts) %in% "ENSMUSG00000002985", 1:10]: object of type 'closure' is not subsettable
~~~
{: .error}
Let's visualize the expression of mouse and human APOE gene across all samples using ggplot2 package.

### Validation of mouse strains

#First we convert the dataframe to longer format and join our covariates by MouseID

~~~
count_tpose <- counts  %>%
                rownames_to_column(.,var="gene_id") %>% 
                filter(gene_id %in% c("ENSG00000130203","ENSMUSG00000002985")) %>% 
                pivot_longer(.,cols = -"gene_id",names_to = "Mouse.ID",values_to="counts") %>% as.data.frame() %>%
                left_join(ind_meta %>% mutate("Mouse.ID"=as.character(Mouse.ID)),by="Mouse.ID") %>% as.data.frame()
~~~
{: .language-r}



~~~
Error in h(simpleError(msg, call)): error in evaluating the argument 'x' in selecting a method for function 'as.data.frame': error in evaluating the argument 'x' in selecting a method for function 'as.data.frame': is.data.frame(df) is not TRUE
~~~
{: .error}



~~~
head(count_tpose) 
~~~
{: .language-r}



~~~
Error in h(simpleError(msg, call)): error in evaluating the argument 'x' in selecting a method for function 'head': object 'count_tpose' not found
~~~
{: .error}



~~~
#make the age column a factor and re-order the levels
count_tpose$Age <- factor(count_tpose$Age,levels=c(4,8,12))
~~~
{: .language-r}



~~~
Error in eval(expr, envir, enclos): object 'count_tpose' not found
~~~
{: .error}



~~~
# rename the gene id to gene symbol
count_tpose$gene_id[count_tpose$gene_id %in% "ENSG00000130203"] <- "Human APOE"
~~~
{: .language-r}



~~~
Error in eval(ei, envir): object 'count_tpose' not found
~~~
{: .error}



~~~
count_tpose$gene_id[count_tpose$gene_id %in% "ENSMUSG00000002985"] <- "Mouse Apoe"
~~~
{: .language-r}



~~~
Error in eval(ei, envir): object 'count_tpose' not found
~~~
{: .error}



~~~
#Create simple box plots showing normalized counts by genotype and time point, faceted by sex.
count_tpose %>% 
  ggplot(aes(x=Age, y=counts, color=Genotype)) +
  geom_boxplot() + 
  geom_point(position=position_jitterdodge()) +
  facet_wrap(~Sex+gene_id) +theme_bw()
~~~
{: .language-r}



~~~
Error in eval(expr, envir, enclos): object 'count_tpose' not found
~~~
{: .error}

You will notice expression of mouse Apoe is higher in samples carrying humanized APOE gene sequences but lower in other samples and vice-versa for human APOE. This also suggest that mouse Apoe gene has been correctly replaced by human APOE gene sequences in APOE4 and APOE4Trem strains. This way we can validated the insertion of human transgene.

As before, we need to merge expression of human APOE and mouse Apoe:

~~~
counts[rownames(counts) %in% "ENSMUSG00000002985",] <- counts[rownames(counts) %in% "ENSMUSG00000002985",] + counts[rownames(counts) %in% "ENSG00000130203",]
~~~
{: .language-r}



~~~
Error in counts[rownames(counts) %in% "ENSMUSG00000002985", ]: object of type 'closure' is not subsettable
~~~
{: .error}



~~~
counts <- counts[!rownames(counts) %in% c("ENSG00000130203"),]
~~~
{: .language-r}



~~~
Error in counts[!rownames(counts) %in% c("ENSG00000130203"), ]: object of type 'closure' is not subsettable
~~~
{: .error}

Let's verify

~~~
counts[,1:6] %>% 
  filter(!str_detect(rownames(.), "MUS"))
~~~
{: .language-r}



~~~
Error in counts[, 1:6]: object of type 'closure' is not subsettable
~~~
{: .error}

Formatting the count and metadata for differential analysis

~~~
rawdata <- counts[,sort(colnames(counts))]
~~~
{: .language-r}



~~~
Error in counts[, sort(colnames(counts))]: object of type 'closure' is not subsettable
~~~
{: .error}



~~~
metadata <- covar[sort(rownames(covar)),] 
~~~
{: .language-r}



~~~
Error in eval(expr, envir, enclos): object 'covar' not found
~~~
{: .error}



~~~
## renaming column names
colnames(metadata) <- c("sex","genotype","timepoint","Batch")
~~~
{: .language-r}



~~~
Error in `colnames<-`(`*tmp*`, value = c("sex", "genotype", "timepoint", : attempt to set 'colnames' on an object with less than two dimensions
~~~
{: .error}



~~~
metadata$sex[metadata$sex %in% "F"] <- "female"
~~~
{: .language-r}



~~~
Error in `*tmp*`$sex: object of type 'closure' is not subsettable
~~~
{: .error}



~~~
metadata$sex[metadata$sex %in% "M"] <- "male"
~~~
{: .language-r}



~~~
Error in `*tmp*`$sex: object of type 'closure' is not subsettable
~~~
{: .error}


## Differential Analysis of all groups in LOAD1 cohort

Next, we create a comparison table that has all cases and controls that we would like to compare with each other. Here I have made comparison group for age and sex-matched distinct mouse strains vs C57BL/6J (control mice):

~~~
metadata$Group <- paste0(metadata$genotype,"-",metadata$sex,"-",metadata$timepoint)
~~~
{: .language-r}



~~~
Error in metadata$genotype: object of type 'closure' is not subsettable
~~~
{: .error}



~~~
comparisons <-  data.frame(
                      control=rep(c("C57BL/6J-male-4", "C57BL/6J-female-4", "C57BL/6J-male-8", "C57BL/6J-female-8", "C57BL/6J-male-12", "C57BL/6J-female-12"),3), 
                              case=c("APOE4-male-4", "APOE4-female-4", "APOE4-male-8", "APOE4-female-8", "APOE4-male-12", "APOE4-female-12",
                                  "Trem2-male-4", "Trem2-female-4", "Trem2-male-8", "Trem2-female-8", "Trem2-male-12", "Trem2-female-12",
                                  "APOE4Trem2-male-4", "APOE4Trem2-female-4", "APOE4Trem2-male-8", "APOE4Trem2-female-8", "APOE4Trem2-male-12", "APOE4Trem2-female-12")
                      )
head(comparisons)
~~~
{: .language-r}



~~~
             control            case
1    C57BL/6J-male-4    APOE4-male-4
2  C57BL/6J-female-4  APOE4-female-4
3    C57BL/6J-male-8    APOE4-male-8
4  C57BL/6J-female-8  APOE4-female-8
5   C57BL/6J-male-12   APOE4-male-12
6 C57BL/6J-female-12 APOE4-female-12
~~~
{: .output}


Finally, we implement our DEG function on each comparison and store the result table in a list and data frame:

~~~
DE_LOAD1.list <- list()
DE_LOAD1.df <- data.frame()

for (i in 1:nrow(comparisons)){
  meta <- metadata[metadata$Group %in% comparisons[i,],]
  DEG(rawdata,meta,ref = "C57BL/6J")
  
  #append results in data frame
  DE_LOAD1.df <- rbind(DE_LOAD1.df,All_res %>% mutate(model=gsub("-.*$","",comparisons[i,2])[1],sex=unique(meta$sex),age=unique(meta$timepoint)))
  
  #append results in list
  DE_LOAD1.list[[i]] <- All_res
  names(DE_LOAD1.list)[i] <- paste0(comparisons[i,2])
}
~~~
{: .language-r}



~~~
Error in h(simpleError(msg, call)): error in evaluating the argument 'i' in selecting a method for function '[': error in evaluating the argument 'x' in selecting a method for function '%in%': object of type 'closure' is not subsettable
~~~
{: .error}



~~~
save(DE_LOAD1.df,DE_LOAD1.list,file="../result/DEAnalysis_LOAD1.Rdata")
~~~
{: .language-r}



~~~
Error in gzfile(file, "wb"): cannot open the connection
~~~
{: .error}

check number of genes significantly expressed (adjusted p =0.05) in all cases compared to age and sex-matched controls:


~~~
degs.up <- map(DE_LOAD1.list, ~length(which(.x$padj<0.05 & .x$log2FoldChange>0)))
degs.down <- map(DE_LOAD1.list, ~length(which(.x$padj<0.05 & .x$log2FoldChange<0)))
deg <- data.frame(comparison=names(degs.up), Up_DEGs.pval.05=as.vector(unlist(degs.up)),Down_DEGs.pval.05=as.vector(unlist(degs.down)))
knitr::kable(deg)
~~~
{: .language-r}

> ## Challenge 5
  > ## Challenge 5
  > Save the data for next lesson?
  >
  > > ## Solution to Challenge 5
  > > save(DE_LOAD1.df,DE_LOAD1.list,file="../result/DEAnalysis_LOAD1.Rdata")
> > 
  > {: .solution}
{: .challenge}


### Session Info

~~~
sessionInfo()
~~~
{: .language-r}



~~~
R version 4.3.0 (2023-04-21)
Platform: aarch64-apple-darwin20 (64-bit)
Running under: macOS Monterey 12.6.2

Matrix products: default
BLAS:   /Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/lib/libRblas.0.dylib 
LAPACK: /Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.11.0

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

time zone: America/New_York
tzcode source: internal

attached base packages:
[1] stats4    stats     graphics  grDevices utils     datasets  methods  
[8] base     

other attached packages:
 [1] clusterProfiler_4.8.1       lubridate_1.9.2            
 [3] forcats_1.0.0               stringr_1.5.0              
 [5] dplyr_1.1.2                 purrr_1.0.1                
 [7] readr_2.1.4                 tidyr_1.3.0                
 [9] tibble_3.2.1                tidyverse_2.0.0            
[11] EnhancedVolcano_1.18.0      ggrepel_0.9.3              
[13] GO.db_3.17.0                org.Mm.eg.db_3.17.0        
[15] AnnotationDbi_1.62.1        ggplot2_3.4.2              
[17] DESeq2_1.40.1               SummarizedExperiment_1.30.2
[19] Biobase_2.60.0              MatrixGenerics_1.12.1      
[21] matrixStats_1.0.0           GenomicRanges_1.52.0       
[23] GenomeInfoDb_1.36.0         IRanges_2.34.0             
[25] S4Vectors_0.38.1            BiocGenerics_0.46.0        
[27] knitr_1.43                 

loaded via a namespace (and not attached):
 [1] DBI_1.1.3               bitops_1.0-7            gson_0.1.0             
 [4] shadowtext_0.1.2        gridExtra_2.3           rlang_1.1.1            
 [7] magrittr_2.0.3          DOSE_3.26.1             compiler_4.3.0         
[10] RSQLite_2.3.1           png_0.1-8               vctrs_0.6.2            
[13] reshape2_1.4.4          pkgconfig_2.0.3         crayon_1.5.2           
[16] fastmap_1.1.1           XVector_0.40.0          ggraph_2.1.0           
[19] utf8_1.2.3              HDO.db_0.99.1           tzdb_0.4.0             
[22] enrichplot_1.20.0       bit_4.0.5               xfun_0.39              
[25] zlibbioc_1.46.0         cachem_1.0.8            aplot_0.1.10           
[28] jsonlite_1.8.5          blob_1.2.4              DelayedArray_0.26.3    
[31] BiocParallel_1.34.2     tweenr_2.0.2            parallel_4.3.0         
[34] R6_2.5.1                RColorBrewer_1.1-3      stringi_1.7.12         
[37] GOSemSim_2.26.0         Rcpp_1.0.10             downloader_0.4         
[40] Matrix_1.5-4            splines_4.3.0           igraph_1.4.3           
[43] timechange_0.2.0        tidyselect_1.2.0        viridis_0.6.3          
[46] qvalue_2.32.0           codetools_0.2-19        lattice_0.21-8         
[49] plyr_1.8.8              treeio_1.24.1           withr_2.5.0            
[52] KEGGREST_1.40.0         evaluate_0.21           gridGraphics_0.5-1     
[55] scatterpie_0.2.1        polyclip_1.10-4         Biostrings_2.68.1      
[58] ggtree_3.8.0            pillar_1.9.0            ggfun_0.0.9            
[61] generics_0.1.3          RCurl_1.98-1.12         hms_1.1.3              
[64] tidytree_0.4.2          munsell_0.5.0           scales_1.2.1           
[67] glue_1.6.2              lazyeval_0.2.2          tools_4.3.0            
[70] data.table_1.14.8       fgsea_1.26.0            locfit_1.5-9.7         
[73] graphlayouts_1.0.0      fastmatch_1.1-3         tidygraph_1.2.3        
[76] cowplot_1.1.1           grid_4.3.0              ape_5.7-1              
[79] colorspace_2.1-0        nlme_3.1-162            patchwork_1.1.2        
[82] GenomeInfoDbData_1.2.10 ggforce_0.4.1           cli_3.6.1              
[85] fansi_1.0.4             viridisLite_0.4.2       S4Arrays_1.0.4         
[88] gtable_0.3.3            yulab.utils_0.0.6       digest_0.6.31          
[91] ggplotify_0.1.0         farver_2.1.1            memoise_2.0.1          
[94] lifecycle_1.0.3         httr_1.4.6              bit64_4.0.5            
[97] MASS_7.3-58.4          
~~~
{: .output}

